test/TEST-64-UDEV-STORAGE/meson.build

```diff
@@ -11,31 +11,253 @@ unit = configure_file(
+
+foreach i : range(4)
+        id = f'drivelvmbasic@i@'
+        cmdline += [f'--drive=@id@:32M::cache=unsafe']
+        qemu_args += ['-device', f'scsi-hd,drive=@id@,vendor=systemd,product=foobar,serial=deadbeeflvm@i@']
+endforeach
+
+udev_storage_tests += {
+        'name' : 'lvm_basic',
+        'cmdline' : cmdline,
+        'qemu-args' : qemu_args,
+}
+
+cmdline = []
+qemu_args = ['-device', 'virtio-scsi-pci,id=scsi0']
+
+foreach i : range(5)
+        id = f'drivemdadmbasic@i@'
+        cmdline += [f'--drive=@id@:64M::cache=unsafe']
+        qemu_args += ['-device', f'scsi-hd,drive=@id@,vendor=systemd,product=foobar,serial=deadbeefmdadm@i@']
+endforeach
+
+udev_storage_tests += {
+        'name' : 'mdadm_basic',
+        'cmdline' : cmdline,
+        'qemu-args' : qemu_args,
+}
+
+cmdline = []
+qemu_args = ['-device', 'virtio-scsi-pci,id=scsi0']
+
+foreach i : range(5)
+        id = f'drivemdadmlvm@i@'
+        cmdline += [f'--drive=@id@:64M::cache=unsafe']
+        qemu_args += ['-device', f'scsi-hd,drive=@id@,vendor=systemd,product=foobar,serial=deadbeefmdadmlvm@i@']
+endforeach
+
+udev_storage_tests += {
+        'name' : 'mdadm_lvm',
+        'cmdline' : cmdline,
+        'qemu-args' : qemu_args,
+}
+
+cmdline = []
+qemu_args = []
+
+# Add 16 multipath devices, each backed by 4 paths
+foreach ndisk : range(16)
+        ndiskfmt = ndisk >= 10 ? f'00@ndisk@' : f'000@ndisk@'
+        wwn = f'0xDEADDEADBEEF@ndiskfmt@'
+        size = ndisk == 0 ? '16M' : '1M'
+
+        foreach nback : range(4)
+                id = f'drive@ndisk@x@nback@'
+                cmdline += [f'--drive=@id@:@size@::cache=unsafe:@ndisk@']
+                qemu_args += ['-device', f'scsi-hd,drive=@id@,serial=MPIO@ndisk@,wwn=@wwn@']
+        endforeach
+endforeach
+
+udev_storage_tests += {
+        'name' : 'multipath_basic_failover',
+        'cmdline' : cmdline,
+        'qemu-args' : qemu_args,
+}
+
+cmdline = []
+qemu_args = []
+
+foreach i : range(10)
+        id = f'drivesimultaneousevents@i@'
+        cmdline += [f'--drive=@id@:128M::cache=unsafe']
+        qemu_args += ['-device', f'scsi-hd,drive=@id@,serial=deadbeeftest@i@']
+endforeach
+
+udev_storage_tests += {
+        'name' : 'simultaneous_events',
+        'cmdline' : cmdline,
+        'qemu-args' : qemu_args,
+}
+
+cmdline = []
+qemu_args = []
+
+foreach i : range(4)
+        qemu_args += ['-device', f'virtio-scsi-pci,id=scsi@i@']
+endforeach
+
+foreach i : range(128)
+        id = f'drive@i@'
+        cmdline += [f'--drive=@id@:1M::cache=unsafe']
+        div = i / 32
+        mod = i % 32
+        qemu_args += ['-device', f'scsi-hd,drive=@id@,bus=scsi@div@.0,channel=0,scsi-id=@mod@,lun=0']
+endforeach
+
+udev_storage_tests += {
+        'name' : 'virtio_scsi_basic',
+        'cmdline' : cmdline,
+        'qemu-args' : qemu_args,
+}
+
+qemu = find_program('qemu-system-@0@'.format(host_machine.cpu_family()), 'qemu-kvm', dirs : ['/usr/libexec'], native : true, required : false)
```

Comment:
Every additional shelling out makes the meson run slower - given these tests don't run normally, can this be behind the flag that they use to be enabled?

Reply:
No, we always make these available, even if they don't run by default, so we have to always configure them properly. If you want to require specifying a meson option to enable these let's discuss that in a separate PR.

I've made sure this stuff if skipped now if want_tests is "false" since that was actually something that was missing.